torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -9,13 +9,28 @@
 
 
 def _orthogonalize(matrix, epsilon=0):
+    """
+    Decide between Gram-Schmidt or QR factorization to orthogonalize the matrix.
+    QR factorization doesn't work with half-precision, but it is usually faster with a rank > 2.
+    """
+    rank = matrix.shape[1]
+    dtype = matrix.dtype
+    if rank <= 2 or dtype in [torch.float16, torch.bfloat16]:
+        _orthogonalize_gram_schmidt(matrix, epsilon=epsilon)
+    else:
+        torch.linalg.qr(
+            matrix,
+            out=(
+                matrix,
+                torch.empty(rank, rank, device=matrix.device, dtype=dtype)
+            )
```

Comment:
This is correct only when `matrix.shape[-2] >= matrix.shape[-1]`.
`torch.linalg.qr` returns two tensors of shapes `(m, k), (k, n)`, where `k=min(m, n)`, `m=matrix.shape[-2]`, and `n=matrix.shape[-1]`

Reply:
That's a very good catch. This shouldn't be an issue for PowerSGD, though. The inputs are of shape `(m, d)` or `(n, d)`, where `d` is defined as `min(n, m, state.matrix_approximation_rank)`. Maybe we can add an assertion for this in the orthogonalize function?